### ISSUE TITLE
Core/localized extensions code duplication?

### ISSUE THREAD

**Issue by krsriq** [Nyholm/next-faker#14]
Do I understand it correctly that under the current proposal, localized extensions would basically have to duplicate what they do when using protected properties, so `$faker->ext(PersonDe::class)->firstName()` would be a copy of what `$faker->ext(PersonEn::class)->firstName()` does (i.e. select a random element from `static::$firstNameFormat`)? 

If so, I guess that's fine for trivial stuff, but how about more advanced cases such as `Provider\Text::realText()`,  `Provider\Payment::iban()` or even `Provider\Internet::userName()`?

**Comment by Nyholm** [comment]
Yes, that is correct. To be clear, the logic we are talking about is: 

```php
public function firstName(): string
{
  return $this->names[array_rand($this->names, 1)];
}
```

For "advanced stuff", Im not sure. There might be some helper methods if we really think we need it.

**Comment by krsriq** [comment]
If we don't use some well-defined way of sharing code between extensions/core, there would be quite a lot of duplication I'm afraid - and with that the increased possibility of bugs in localized extensions, maintenance overhead etc.

I mentioned [`realText`](https://github.com/FakerPHP/Faker/blob/23019f8faffd9a837f7e20f1d56a6108180cb894/src/Faker/Provider/Text.php#L29) as one of the "advanced stuff" cases - there's quite a lot happening there in the core provider, I think we really wouldn't want to replicate all that in each localized provider, no? 

And right now we have an easy way for localized providers to use (and override some of the) methods `realText` uses, such as is done in [`zh_TW`](https://github.com/FakerPHP/Faker/blob/23019f8faffd9a837f7e20f1d56a6108180cb894/src/Faker/Provider/zh_TW/Text.php#L803) or in [`ar_JO\Text`](https://github.com/FakerPHP/Faker/blob/23019f8faffd9a837f7e20f1d56a6108180cb894/src/Faker/Provider/ar_JO/Text.php#L7). How could  we achieve a similar functionality for Faker 2.0?

**Comment by Nyholm** [comment]
> I think we really wouldn't want to replicate all that in each localized provider, no?

Correct. 

> How could we achieve a similar functionality for Faker 2.0?

Im not sure. There might be some helper methods if we really think we need it.

Do you have any more concrete suggestions?

**Comment by krsriq** [comment]
One solution would be to keep using the template method pattern, as is (more or less) done with the current `Provider\Text`, i.e. by having an abstract `FileExtension` class in the core, which is extended by the localized extensions (e.g. `English\Text`). This obviously does not preclude adding a `TextContract` interface (as you have done in your proposal with `Extension\File`).

The advantages of this approach are apparent with what is done with the current `Provider\Text`: flexibility to add language-specific data and overrides for some methods, while re-using the rest. 

I couldn't think of any obvious other solutions that only use 'helper methods' to achieve similar functionality (without adding too much complexity, and ideally while favoring composition), though I'd definitely be happy to learn how to do this differently :)

**Comment by Nyholm** [comment]
Yeah.. I really would like to avoid inheritance. It is one of the reasons the current architecture is a large mess. 

I just started to add some helpers in: https://github.com/FakerPHP/Faker/pull/162
I'll keep building on the Swedish one to see exactly what problems I run in to. Maybe I dont know enough of Faker to see the big problems.
